### ISSUE TITLE
Requirements: Boolean operators on rules

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#773]
There should be and/or/not operations on rules, which have other rules as subfields.

**Comment by cdsmith** [comment]
The tricky part here is messages on failure: "and" is useless on its own, since a requirement already has a list of rules which are and'ed together; but "or" and "not" raise questions about how to describe why they failed.  As prior art, the gMock testing library for C++ contains a fairly complex system of template-based explainers, which knows whether a statement is in a positive or negative position, etc.  But it still falls short of what's usable for children, when there's complex Boolean logic involved.

It's likely that the ordinary set of Boolean operators isn't enough.  We care about the intent behind the logical relationship.  For example, if you mean "whenever foo is defined, it matches this pattern", then one wants the failure message to talk only about whether it matches that pattern, and not mention its being defined.  But if that's encoded as "either of: (1) foo is undefined, or (2) foo matches the pattern", it's not clear how we're supposed to know which is that dominant clause.

I'll have to think more about this.

**Comment by alphalambda** [comment]
This reflection suggests to me that we need some kind of "combinator" model.

We need to apply several rules in sequence, as long as they succeed,
and we also need to have alternative rule sequences that are tried when the previous alternate fails.

The underlying monad can carry information about why each rule failed.

**Comment by cdsmith** [comment]
I've resolved what I know I want to do here, by adding these rule forms:

- `ifThen`: if (a) matches, then (b) must match.
- `all`: all of these rules must match.
- `any`: at least one of these rules must match.
- `not`: this rule must not match.

To help a bit with errors, I also added:

- `onFailure`: this rule must match, but if it fails, use the given error message instead of the one from the underlying rule.

I played around with this a little.  It works out okay, though it can get wordy at times.  I find myself wanting some kind of abstraction in the rule language... oh well, that's definitely a different feature request!
